text/0000-ptr-to-field.md

```diff
@@ -0,0 +1,395 @@
+
+They will also need to interact with the privacy rules for fields, and will have the same privacy as the field that they are derived from.
+
+You can make a field type for the following types of types (some example syntax is shown)
+* tuples `(T, U, ...)`
+    * `<(T, U)>.0`
+* tuple structs `Foo(T, U, ...)`
+    * `Foo.0`
+* structs `struct Foo { field: Field, ... }`
+    * `Foo.field`
+* unions `union Foo { field: Field }`
+    * same syntax as structs
+    * constructing a field type is `unsafe`, this is because accessing fields of `union`s is unsafe
+
+The compiler can decide whether to actual generate a field type, this is to help compile times (if you don't use field types, then the compiler shouldn't slow down too much because of this feature).
+
+## `*const T`/`*mut T`
+
+`project_unchecked` and `wrapping_project` will both live inside of `core::ptr`.
+
+We need both `project_unchecked` and `wrapping_project` because there are some important optimization available inside of LLVM related to aliasing and escape analysis. In particular the LLVM `inbounds` assertion tells LLVM that a pointer offset stays within the same allocation and if the pointer is invalid or the offset does not stay within the same allocation it is considered UB. This behaviour is exposed via `project_unchecked`. This can be used by implementations of the `Project` trait for smart pointers that are always valid, like `&T` to enable better codegen. `wrapping_project` on the other hand will not assert `inbounds`, and will just wrap around if the pointer offset is larger than `usize::max_value()`. This safe defined behaviour, even if it is almost always a bug, unlike `project_unchecked` which is UB on invalid pointers or offsets.
+
+This corresponds with `core::ptr::add` and `core::ptr::wrapping_add` in safety and behaviour. `project_unchecked` and `project` need to be implemented as intrinsics because there is no way to assert that the pointer metadata for fat pointers of `Field::Parent` and `Field::Type` will always match in general without some other compiler support. This is necessary to allow unsized types to be used transparently with this scheme.
+
+`inverse_project_unchecked` and `inverse_wrapping_project` are just like their counterparts in safety. `inverse_project_unchecked` is UB to use on invalid pointers, where `inverse_wrapping_project` just wraps around on overflow. But there is one important safety check that must be performed before dereferencing the resulting pointer. The resulting pointer may not actually point to a valid `*const F::Parent` if `*const F::Type` does not live inside of a `F::Parent`, so one must take care to ensure that the parent pointer is indeed valid. This is different from `project_unchecked` and `wrapping_project` because there one only needs to validate the original pointer, not the resulting pointer.
```

Comment:
I don't think this is quite right. I think better wording is along the lines of

> `inverse_project_unchecked` is not UB and `inverse_wrapping_project` produces a valid pointer if and only if the initial pointer is both valid and points to a field in the parent type used to perform the inverse projection.

This is all prerequisites on the input pointer to match the field descriptor provided, not something that should (or even could) be checked on the resulting pointer.

This, of course, makes the resulting pointer a lot harder to use safely, as you need more information from outside the type system to make use sound.

Reply:
That wording is much better, thanks! That does look right.